repose-aggregator/components/services/container-configuration-service/impl/src/main/scala/org/openrepose/nodeservice/containerconfiguration/ContainerConfigurationServiceImpl.scala

```diff
@@ -0,0 +1,146 @@
+
+import java.lang.Long
+import java.util.Optional
+import java.util.concurrent.atomic.{AtomicBoolean, AtomicReference}
+import java.util.function.UnaryOperator
+import javax.annotation.{PostConstruct, PreDestroy}
+import javax.inject.{Inject, Named}
+
+import com.typesafe.scalalogging.slf4j.LazyLogging
+import org.openrepose.commons.config.manager.UpdateListener
+import org.openrepose.core.container.config.{ContainerConfiguration, DeploymentConfiguration}
+import org.openrepose.core.services.config.ConfigurationService
+import org.openrepose.core.spring.ReposeSpringProperties
+import org.springframework.beans.factory.annotation.Value
+
+import scala.collection.JavaConversions._
+
+@Named
+class ContainerConfigurationServiceImpl @Inject()(@Value(ReposeSpringProperties.NODE.CLUSTER_ID) clusterId: String,
+                                                  configurationService: ConfigurationService)
+  extends ContainerConfigurationService with UpdateListener[ContainerConfiguration] with LazyLogging {
+
+  import ContainerConfigurationServiceImpl._
+
+  private val initialized = new AtomicBoolean(false)
+  private val patchedDeploymentConfiguration = new AtomicReference[DeploymentConfiguration]()
+  private val updateListeners = new AtomicReference[Set[UpdateListener[DeploymentConfiguration]]](Set.empty)
+
+  @PostConstruct
+  def init(): Unit = {
+    logger.debug("Initializing service and registering configuration listener")
+    val xsdUrl = getClass.getResource("/META-INF/schema/container/container-configuration.xsd")
+
+    configurationService.subscribeTo(
+      ContainerConfigurationFilename,
+      xsdUrl,
+      this,
+      classOf[ContainerConfiguration]
+    )
+  }
+
+  @PreDestroy
+  def destroy(): Unit = {
+    logger.debug("Un-registering configuration listener and shutting down service")
+    configurationService.unsubscribeFrom(ContainerConfigurationFilename, this)
+  }
+
+  override def getVia: Optional[String] = {
+    initializationCheck()
+    Optional.ofNullable(patchedDeploymentConfiguration.get().getVia)
+  }
+
+  override def getContentBodyReadLimit: Optional[Long] = {
+    initializationCheck()
+    Optional.ofNullable(patchedDeploymentConfiguration.get().getContentBodyReadLimit)
+  }
+
+  override def getDeploymentConfiguration: DeploymentConfiguration = {
+    initializationCheck()
+    patchedDeploymentConfiguration.get()
+  }
+
+  override def subscribeTo(listener: UpdateListener[DeploymentConfiguration]): Unit = {
+    subscribeTo(listener, sendNotificationNow = true)
+  }
+
+  override def subscribeTo(listener: UpdateListener[DeploymentConfiguration], sendNotificationNow: Boolean): Unit = {
+    initializationCheck()
+    logger.debug("Subscribing listener with hash code: {}", listener.hashCode.toString)
+
+    // TODO: Clean this up with Scala 2.12 support for Scala function -> Java 8 function
+    updateListeners.updateAndGet(new UnaryOperator[Set[UpdateListener[DeploymentConfiguration]]] {
```

Comment:
Well, it will mostly get cleaned up when we move to Scala 2.12 and do this TODO.

Still, dropping the atomic references would clean things up. I'm not _that_ concerned with threading issues, but they could happen. But what's the chance of more than one configuration update being called at the same time? Or of subscription and unsubscription between called for the same listener across threads?

Open for suggestions here, as well.